server/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java

```diff
@@ -162,27 +162,45 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request, Task task)
 
     public static class Request extends BaseNodesRequest<Request> {
 
-        private ShardId shardId;
+        private final ShardId shardId;
+        @Nullable
+        private final String customDataPath;
```

Comment:
Maybe we should add a short sentence of javadoc here or to `Request` itself explaining why we send the `customDataPath` in this request? It's going to be non-trivial to understand this without going back to this PR for future generations I think :)

Reply:
I've added some sentences in 3bd2843979a. I think this will all look simpler once we can remove the BWC path. Then it's going to be very simple. We just lookup by whatever we got in the request, no need to look at local cluster state.